File: Problem/By section/Divide and Conquer/Problem 241/problem_241.java
import java.util.*;

class Solution {
    // memory for alreday calculated substring result 
    private HashMap<String, List<Integer>> map = new HashMap<>();
    
    // Time: O(w(1) * w(n - 1) + w(2) * w(n - 2) + ... + w(n - 1) * w(1))
    // Space: Same
    public List<Integer> diffWaysToCompute(String input) {
        // if we already calculate this expression
        // use it from the hashmap
        if(map.containsKey(input)) return map.get(input);
        
        // if not
        List<Integer> curr_ans = new ArrayList<>(); 
        for(int i = 0; i < input.length(); i ++){
            char temp = input.charAt(i);
            // we skip all the number till we get an operation
            if(temp == '+' || temp == '-' || temp == '*' ){
                
                // we enmuarate all the possible outcome from the leftside and rightside of the opearation 
                List<Integer> left = diffWaysToCompute(input.substring(0, i));
                List<Integer> right = diffWaysToCompute(input.substring(i + 1, input.length()));

                // then we combine these two subset to a new huge set
                // according to the opearator between them
                for(int l : left){
                    for(int r : right){
                        switch(temp){
                            case '+':
                                curr_ans.add(l + r);
                                break;
                            case '-':
                                curr_ans.add(l - r);
                                break;
                            case '*':
                                curr_ans.add(l * r);
                                break;
                        }
                    }
                }
            }
        }
        
        // important here!!!!
        // if we did not meet any opearator
        // meaning there is only number here
        curr_ans = (curr_ans.size() == 0) ? Arrays.asList(Integer.parseInt(input)) : curr_ans;

        // record the susbet for this input
        map.put(input, curr_ans);
        
        return curr_ans;
    }
}

File: Problem/By section/Divide and Conquer/Problem 241/problem_241.cpp
#include <bits/stdc++.h>

using namespace std;

void print_vec(vector<int>& vec);

class Solution {
public:
    unordered_map<string, vector<int>> reminder;

    vector<int> diffWaysToCompute(string expression) {
        if(expression.length() == 0) {
            return vector<int>{0};
        }if(reminder.count(expression)) {
            return reminder[expression];
        }else{
            vector<int> result;
            bool meet_operator = false;

            for(auto i = begin(expression); i != end(expression); i++) {
                if(*i == '+' || *i == '-' || *i == '*' || *i == '/') {
                    meet_operator = true;
                    vector<int> left = diffWaysToCompute(string(begin(expression), i));
                    vector<int> right = diffWaysToCompute(string(i + 1, end(expression)));

                    calculate(left, right, result, *i);
                }
            }

            result = (!meet_operator) ? vector<int>{stoi(expression)} : result;
            reminder[expression] = result;

            return result;
        }
    }

    void calculate(vector<int>& left, vector<int>& right, vector<int>& result, char expression) {
        for(auto& i : left) {
            for(auto& j : right) {
                int single_result;

                switch(expression) {
                    case '+':
                        single_result = i + j;
                        break;
                    case '-':
                        single_result = i - j;
                        break;
                    case '*':
                        single_result = i * j;
                        break;
                    case '/':
                        single_result = i / j;
                        break;
                }

                result.push_back(single_result);
            }
        }
    }
};

void print_vec(vector<int>& vec) {
    for(auto i : vec) {
        cout << i << " ";
    }

    cout << endl;
}

int main(int argc, char** argv) {
    Solution s;
    char test = 'a';
    string a = "11";
    auto result = s.diffWaysToCompute(a);
    print_vec(result);
    return 0;
}